operators/pkg/controller/elasticsearch/driver/downscale.go

```diff
@@ -60,9 +39,15 @@ func HandleDownscale(
 		return results.WithError(err)
 	}
 
+	// make sure we only downscale nodes we're allowed to
+	invariants, err := NewDownscaleInvariants(downscaleCtx.k8sClient, downscaleCtx.es)
```

Comment:
This got me wondering about the overall signatures of the downscale related funcs. At first it seemed a little odd that this func consumes a k8s client and an ES client, but then its caller consumes both of those clients, as well as the ssets it will operate against (rather than querying for the actual ssets itself). But in this case `NewDownscaleInvariants()` only needs the k8s client to get the list of pods. I don't really have a preference or a suggestion here, just making an observation that choosing the right level of abstraction can be weird. It's a little tempting to just pass this a `downscaleContext`, but again I'm mostly just thinking out loud. The trade offs either direction here are minimal at best so I wouldn't sweat it.

Reply:
Maybe moving to less responsibility in the invariants constructor here would make things cleaner:
`NewDownscaleInvariants(runningMasters int)`
where we would compute runningMasters beforehand.

Reply:
I think it looks better in https://github.com/elastic/cloud-on-k8s/pull/1549/commits/8b59ee220f00695abe499aa7e2e9a329168c08cc where state is decoupled form invariants checks.